planner/core/stats.go

```diff
@@ -146,9 +147,114 @@ func (ds *DataSource) DeriveStats(childStats []*property.StatsInfo) (*property.S
 			break
 		}
 	}
+	// consider the IndexMergePath. Now, we just generate `IndexMergePath` in DNF case.
+	if len(ds.pushedDownConds) > 0 && len(ds.possibleAccessPaths) > 1 && ds.ctx.GetSessionVars().EnableIndexMerge {
+		needConsiderIndexMerge := true
+		for i := 1; i < len(ds.possibleAccessPaths); i++ {
+			if len(ds.possibleAccessPaths[i].accessConds) != 0 {
+				needConsiderIndexMerge = false
+				break
+			}
+		}
+		if needConsiderIndexMerge {
+			ds.generateIndexMergeOrPaths()
+		}
+	}
 	return ds.stats, nil
 }
 
+// getIndexMergeOrPath generates all possible IndexMergeOrPaths.
+func (ds *DataSource) generateIndexMergeOrPaths() {
+	usedIndexCount := len(ds.possibleAccessPaths)
+	for i, cond := range ds.pushedDownConds {
+		var partialPaths = make([]*accessPath, 0, usedIndexCount)
+
+		sf, ok := cond.(*expression.ScalarFunction)
+		if !ok || sf.FuncName.L != ast.LogicOr {
+			continue
+		}
+		dnfItems := expression.FlattenDNFConditions(sf)
+		for _, item := range dnfItems {
+			var itemPaths []*accessPath
+			cnfItems := expression.SplitCNFItems(item)
+			itemPaths = ds.accessPathsForConds(cnfItems, usedIndexCount)
+			if len(itemPaths) == 0 {
+				partialPaths = nil
+				break
+			}
+			partialPath := ds.buildIndexMergePartialPath(itemPaths)
+			if partialPath == nil {
+				partialPaths = nil
+				break
+			}
+			partialPaths = append(partialPaths, partialPath)
+		}
+		if len(partialPaths) > 1 {
+			possiblePath := ds.buildIndexMergeOrPath(partialPaths, i)
+			if possiblePath != nil {
+				ds.possibleAccessPaths = append(ds.possibleAccessPaths, possiblePath)
+			}
+		}
+	}
+}
+
+// accessPathsForConds generates all possible index paths for conditions.
+func (ds *DataSource) accessPathsForConds(conditions []expression.Expression, usedIndexCount int) []*accessPath {
+	var results = make([]*accessPath, 0, usedIndexCount)
+	for i := 0; i < usedIndexCount; i++ {
+		path := &accessPath{}
+		var err error
+		if ds.possibleAccessPaths[i].isTablePath {
+			path.isTablePath = true
+			_, err = ds.deriveTablePathStats(path, conditions)
+		} else {
+			// if accessConds is empty or tableFilter is not empty, we ignore the access path.
+			// now these conditions are too strict.
+			// for example, a sql `select * from t where a > 1 or (b < 2 and c >3)` and table `t` with indexes on a and b separately.
+			// we can generate a `IndexMergePath` with table filter `a > 1 or (b < 2 and c >3)`.
+			// TODO: solve the above case
+			path.index = ds.possibleAccessPaths[i].index
+			_, err = ds.deriveIndexPathStats(path, conditions)
+		}
+		if err != nil {
+			return results
+		}
+		if len(path.tableFilters) > 0 || len(path.accessConds) == 0 {
+			continue
+		}
+		results = append(results, path)
+
+	}
+	return results
+}
+
+// buildIndexMergePartialPath chooses the best index path from all possible paths.
+// Now we just choose the index with most columns.
```

Comment:
Add a TODO here to indicate that we should improve this strategy, because it is not always better to choose index with most columns, e.g, filter is `c > 1` and the input indexes are `c` and `c_d_e`, the former one is enough, and it is less expensive in execution compared with the latter one.